flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/memory/MemoryTierProducerAgent.java

```diff
@@ -0,0 +1,175 @@
+        this.isBroadcastOnly = isBroadcastOnly;
+        this.memoryManager = memoryManager;
+        this.numSubpartitionEmitBytes = new int[numSubpartitions];
+        this.nettyServiceRegistered = new boolean[numSubpartitions];
+        this.subpartitionProducerAgents = new MemoryTierSubpartitionProducerAgent[numSubpartitions];
+
+        Arrays.fill(numSubpartitionEmitBytes, 0);
+        nettyService.registerProducer(partitionId, this);
+        for (int subpartitionId = 0; subpartitionId < numSubpartitions; ++subpartitionId) {
+            subpartitionProducerAgents[subpartitionId] =
+                    new MemoryTierSubpartitionProducerAgent(subpartitionId, nettyService);
+        }
+        resourceRegistry.registerResource(partitionId, this::releaseResources);
+    }
+
+    @Override
+    public boolean tryStartNewSegment(TieredStorageSubpartitionId subpartitionId, int segmentId) {
+        if (isBroadcastOnly) {
+            return false;
+        }
+        boolean canStartNewSegment =
+                isSubpartitionRegistered(subpartitionId)
+                        && (memoryManager.getMaxNonReclaimableBuffers(this)
+                                        - memoryManager.numOwnerRequestedBuffer(this))
+                                > segmentSizeBuffers;
+        if (canStartNewSegment) {
+            getSubpartitionProducerAgent(subpartitionId.getSubpartitionId())
+                    .addSegmentId(segmentId);
+        }
+        return canStartNewSegment;
+    }
+
+    @Override
+    public boolean tryWrite(TieredStorageSubpartitionId subpartitionId, Buffer finishedBuffer) {
+        int subpartitionIndex = subpartitionId.getSubpartitionId();
+        if (numSubpartitionEmitBytes[subpartitionIndex] != 0
+                && numSubpartitionEmitBytes[subpartitionIndex] + finishedBuffer.readableBytes()
+                        > segmentSizeBytes) {
+            appendEndOfSegmentEvent(subpartitionIndex);
+            numSubpartitionEmitBytes[subpartitionIndex] = 0;
+            return false;
+        }
+        numSubpartitionEmitBytes[subpartitionIndex] += finishedBuffer.readableBytes();
+        append(finishedBuffer, subpartitionIndex);
+        return true;
+    }
+
+    @Override
+    public void connectionEstablished(
+            TieredStorageSubpartitionId subpartitionId,
+            NettyConnectionWriter nettyConnectionWriter) {
+        if (isBroadcastOnly) {
+            throw new RuntimeException(
+                    "Broadcast-only result partition is not allowed to use the memory tier.");
+        }
+        this.subpartitionProducerAgents[subpartitionId.getSubpartitionId()].registerNettyService(
+                nettyConnectionWriter);
+        nettyServiceRegistered[subpartitionId.getSubpartitionId()] = true;
+    }
+
+    @Override
+    public void connectionBroken(NettyConnectionId connectionId) {
+        // noop
+    }
+
+    @Override
+    public void close() {
+        // noop
```

Comment:
I think buffers in memory tier should be released at the closing time, not release resources. The former means the upstream task has finished, while the latter means the result partition can be released.

Reply:
Added comments about this to describe the reason why we do not wait for the netty queue to be consumed when closing the agent.